src/core/AutomationPattern.cpp

```diff
@@ -329,46 +505,62 @@ void AutomationPattern::applyDragValue()
 
 float AutomationPattern::valueAt( const TimePos & _time ) const
 {
+	QMutexLocker m(&m_patternMutex);
+
 	if( m_timeMap.isEmpty() )
 	{
 		return 0;
 	}
 
-	if( m_timeMap.contains( _time ) )
+	// If we have a node at that time, just return its value
+	if (m_timeMap.contains(_time))
 	{
-		return m_timeMap[_time];
+		// When the time is exactly the node's time, we want the inValue
+		return m_timeMap[_time].getInValue();
 	}
 
-	// lowerBound returns next value with greater key, therefore we take
-	// the previous element to get the current value
-	timeMap::ConstIterator v = m_timeMap.lowerBound( _time );
+	// lowerBound returns next value with equal or greater key. Since we already
+	// checked if the key contains a node, we know the returned node has a greater
+	// key than _time. Therefore we take the previous element to calculate the current value
+	timeMap::const_iterator v = m_timeMap.lowerBound(_time);
 
 	if( v == m_timeMap.begin() )
 	{
 		return 0;
 	}
 	if( v == m_timeMap.end() )
 	{
-		return (v-1).value();
+		// When the time is after the last node, we want the outValue of it
+		return OUTVAL(v - 1);
 	}
 
-	return valueAt( v-1, _time - (v-1).key() );
+	return valueAt(v - 1, _time - POS(v - 1));
 }
 
 
 
 
+// This method will get the value at an offset from a node, so we use the outValue of
+// that node and the inValue of the next node for the calculations.
 float AutomationPattern::valueAt( timeMap::const_iterator v, int offset ) const
 {
-	if( m_progressionType == DiscreteProgression || v == m_timeMap.end() )
+	QMutexLocker m(&m_patternMutex);
+
+	// We never use it with offset 0, but doesn't hurt to return a correct
+	// value if we do
+	if (offset == 0) { return INVAL(v); }
+
+	if (m_progressionType == DiscreteProgression)
 	{
-		return v.value();
+		return OUTVAL(v);
 	}
 	else if( m_progressionType == LinearProgression )
 	{
-		float slope = ((v+1).value() - v.value()) /
-							((v+1).key() - v.key());
-		return v.value() + offset * slope;
+		float slope =
+			(INVAL(v + 1) - OUTVAL(v)) /
+			(POS(v + 1) - POS(v));
```

Comment:
IMO it's a little easier to read with a leading operator.
```suggestion
			(INVAL(v + 1) - OUTVAL(v))
			/ (POS(v + 1) - POS(v));
```